pkg/lakom/config/config.go

```diff
@@ -0,0 +1,129 @@
+// SPDX-FileCopyrightText: SAP SE or an SAP affiliate company and Gardener contributors
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package config
+
+import (
+	"crypto"
+	"crypto/rsa"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/gardener/gardener-extension-shoot-lakom-service/pkg/lakom/utils"
+
+	"k8s.io/utils/ptr"
+	"sigs.k8s.io/yaml"
+)
+
+// LoadConfig reads and validates lakom configuration from given file path.
+func LoadConfig(filename string) (*Config, error) {
+	rawConfig, err := os.ReadFile(filepath.Clean(filename))
+	if err != nil {
+		return nil, fmt.Errorf("failed to read config file, %w", err)
+	}
+
+	c := &Config{}
+	if err := yaml.Unmarshal(rawConfig, c); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal config, %w", err)
+	}
+
+	if err := c.validate(); err != nil {
+		return nil, fmt.Errorf("failed to validate config, %w", err)
+	}
+
+	return c, nil
+}
+
+func (c *Config) validate() error {
+	usedNames := map[string]any{}
+
+	for idx, k := range c.PublicKeys {
+		if k.Name == "" {
+			return fmt.Errorf("key at index %d has empty name", idx)
+		}
+
+		if _, ok := usedNames[k.Name]; ok {
+			return fmt.Errorf("duplicated key name %q at index %d", k.Name, idx)
+		}
+		usedNames[k.Name] = nil
+
+		if keys, err := utils.GetCosignPublicKeys([]byte(k.Key)); err != nil {
+			return fmt.Errorf("failed to parse public key for %q (index %d): %w", k.Name, idx, err)
+		} else if len(keys) != 1 {
+			return fmt.Errorf("expected to find exactly one public key, but found %d keys for key name %q (index %d)", len(keys), k.Name, idx)
+		}
+	}
+
+	return nil
+}
+
+// Complete transforms the user exposed config to the internal representation of the config.
+func (c *Config) Complete() (*CompletedConfig, error) {
+	res := CompletedConfig{
+		Keys: []VerifierKey{},
+	}
+
+	for idx, k := range c.PublicKeys {
+		config := VerifierKey{
+			Name: k.Name,
+		}
+
+		keys, err := utils.GetCosignPublicKeys([]byte(k.Key))
+		if err != nil {
+			return nil, err
+		}
+
+		config.Key = keys[0]
+		hash, scheme, err := parseAlgorithm(config.Key, k.Algorithm)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse algorithm for key %q(%d), %w", config.Name, idx, err)
+		}
+
+		config.Hash = hash
+		config.Scheme = scheme
+
+		res.Keys = append(res.Keys, config)
+	}
+
+	return &res, nil
+}
+
+func parseAlgorithm(key crypto.PublicKey, algorithm AlgorithmKey) (*crypto.Hash, *RSASchemeKey, error) {
+	defaultHash := crypto.SHA256
+
+	switch key.(type) {
+	case *rsa.PublicKey:
+		return parseRSAAlgorithm(algorithm)
+	default:
+		return &defaultHash, nil, nil
+	}
+}
+
+func parseRSAAlgorithm(algorithm AlgorithmKey) (*crypto.Hash, *RSASchemeKey, error) {
+	switch algorithm {
+	case "":
+		// RSASSA-PKCS1-v1_5-SHA256 is the default algorithm
+		fallthrough
+
+	// RSASSA-PKCS1-v1_5
+	case RSAPKCS1v15SHA256:
+		return ptr.To(crypto.SHA256), ptr.To(RSAPKCS1v15), nil
```

Comment:
What's the advantage of using `ptr.To` instead of just returning address with `&`?

Also, are the Hash and RSASchemeKey pointers only for optimization purposes?

Reply:
It is not allowed to get the address of a constant, see https://go.dev/ref/spec#Address_operators
```bash
$ make start-lakom
...
# github.com/gardener/gardener-extension-shoot-lakom-service/pkg/lakom/config
pkg/lakom/config/config.go:112:11: invalid operation: cannot take address of crypto.SHA256 (constant 5 of type crypto.Hash)
pkg/lakom/config/config.go:112:27: invalid operation: cannot take address of RSAPKCS1v15 (constant "RSASSA-PKCS1-v1_5" of type RSASchemeKey)
```

Reply:
Thanks! Did not know that. It does make sense.